zebra/zebra_fpm_netlink.c

```diff
@@ -199,12 +203,29 @@ static int netlink_route_info_add_nh(struct netlink_route_info *ri,
 	if (re && CHECK_FLAG(re->flags, ZEBRA_FLAG_EVPN_ROUTE)) {
 		nhi.encap_info.encap_type = FPM_NH_ENCAP_VXLAN;
 
-		zl3vni = zl3vni_from_vrf(nexthop->vrf_id);
-		if (zl3vni && is_l3vni_oper_up(zl3vni)) {
-
-			/* Add VNI to VxLAN encap info */
-			nhi.encap_info.vxlan_encap.vni = zl3vni->vni;
+		/* Extract VNI id for the nexthop SVI interface */
+		zvrf = zebra_vrf_lookup_by_id(nexthop->vrf_id);
+		if (zvrf) {
+			ifp = if_lookup_by_index_per_ns(zvrf->zns,
+							nexthop->ifindex);
+			if (ifp) {
+				zif = (struct zebra_if *)ifp->info;
+				if (zif) {
+					if (IS_ZEBRA_IF_BRIDGE(ifp))
+						link_if = ifp;
+					else if (IS_ZEBRA_IF_VLAN(ifp))
+						link_if =
+						if_lookup_by_index_per_ns(
+							zvrf->zns,
+							zif->link_ifindex);
+					if (link_if)
+						vni = vni_id_from_svi(ifp,
+								      link_if);
+				}
+			}
 		}
+
+		nhi.encap_info.vxlan_encap.vni = vni;
```

Comment:
Do you happen to have a dump of the kernel route entries that you can post here?

Reply:
Kernel route entries:

100.0.0.100 via 50.0.1.22 dev br100 proto 186 metric 20 

Note that above change is only for FPM. This encap_info attribute is only used by the FPM here and FPM encodes this even in the pure type-5 case. It is not needed by the kernel data plane. Kernel should do the routing based on the nexthop interface (br100) and nexthop 50.0.1.22.